### DISCUSSION TITLE
next/link: difference between passing a template string vs passing a URL object to <Link> href prop

### DISCUSSION THREAD

**Discussion by domvo** [vercel/next.js#34061] [Help]
We have a weird problem currently and we can't wrap our head around it. We have a dynamic route in this path: `pages/p/[id].tsx` and we use `getStaticProps` with it that does a very basic fetch request based on the id. So far so good.

We want to link to that route from another page. **This does not work:**

```jsx
// doesn't work
<Link href={`/p/${productId}`}><a>...</a></Link>

// doesn't work either
<Link href={`/p/${encodeURIComponent(productId)}`}><a>...</a></Link>
```

The page does not render correctly, getStaticProps is not called and the page is not hydrated. This happens both with `next dev` and `next start` after building.

The only way we could make it work was like such:

```jsx
<Link
  href={{
    pathname: '/p/[productId]',
    query: { productId },
  }}
>
  <a className="WishlistEntry__Link">{name}</a>
</Link>
```

The markup that gets generated is exactly the same! According to the docs, [both variants](https://nextjs.org/docs/routing/introduction#linking-to-dynamic-paths) should work.

We just would like to understand what's the difference. Or is this maybe a bug that has been introduced recently?

**Comment by robertpaul01** [comment]
Are you using `rewrites` in your `next.config.js`? I was running into this same issue, but found that it was happening because I was using `rewrites` to proxy requests like this:

```
  async rewrites() {
    return [
      {
        source: "/:path*",
        destination: "http://localhost:8080/:path*",
      },
    ];
  }
```

After trying to work around this for a bit, I found that using this as a method to proxy requests is unstable and has a lot of side effects.

**Reply by domvo** [reply]
We found the problem. We had a special setup without a real pages/ folder. The pages folder was a symlink to another folder and apparently this broke something under the hood.

**Comment by domvo** [comment]
Can be closed.
